### PR TITLE
Add operators {not, and, or, lshift, rshift, xor}

### PR DESCRIPTION
**Pull request by csarofeen**
Added operators, fixed some minor things for long int fusion. Still need to enable other int types.

### CODE REVIEW

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -508,6 +537,50 @@ def test_binary_ops(self):
...
+                     "Requires fusion optimization pass to be effective")
+    def test_binary_bitwise(self):
+        def jit_or(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
+            return x & y | z
+
+        def jit_xor(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
+            return x & y ^ z
+
+        def jit_lshift(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
+            return x & y << z
```
Comment (tlemo) [inline test/test_jit_cuda_fuser.py]: what is the precedence here? parentheses to make it unambiguous?

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 2]
```diff
@@ -342,23 +326,29 @@ class CudaKernelGenerator : private kir::IrVisitor {
       code_ << " = ";
     }
 
-    if (auto op = inline_op_str(node->operation())) {
-      code_ << *op << gen(node->in());
+    auto op_type = node->operation();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/codegen.cpp]: const

File: torch/csrc/jit/codegen/cuda/type.cpp [hunk 3]
```diff
@@ -8,6 +8,62 @@ namespace jit {
...
+
+bool isIntegerOp(const BinaryOpType bopt) {
+  return bopt >= BinaryOpType::Mod && bopt <= BinaryOpType::Xor;
+}
+
+bool isLogicalOp(const BinaryOpType bopt) {
+  return bopt >= BinaryOpType::Eq && bopt <= BinaryOpType::NE;
+}
+
+bool maybeBooleanOperator(const BinaryOpType bopt) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/type.cpp]: why 'maybe' instead of 'is'?

**Review by tlemo — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/type.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/type.cpp]: I guess "also" would be a reasonable interpretation, will change to `alsoBooleanOperator` it's strange that PyTorch doesn't have a separate `and` from `bitwise_and` that they actually use.

**Comment by csarofeen** [comment]
messed up the merge hence the forced push

File: test/test_jit_cuda_fuser.py [hunk 4]
```diff
@@ -508,6 +537,50 @@ def test_binary_ops(self):
...
+
+            jitted = torch.jit.script(jit_func)
+            jit_o = jitted(x, y, z)
+            jit_o = jitted(x, y, z)
+            o = jit_func(x, y, z)
+            self.assertEqual(o, jit_o)
+            self.assertGraphContains(jitted.graph_for(x, y, z), FUSION_GUARD)
+
+        # We shouldn't need this redefinition of the function, but otherwise it won't recompile for a new type
+        def jit_or(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:569]: Note to myself, recompilation should work with type guard. But it then becomes tricky to confirm that on python side, so it's ok that we have another definition of the graph.

File: torch/csrc/jit/codegen/cuda/type.h [hunk 5]
```diff
@@ -98,19 +109,40 @@ enum class BinaryOpType {
...
 
-  // Logical Ops
-  // Int operations, leave position of Mod we depend on its location of first
+  // Integer output ops. If changing modify isIntegerOp
   Mod,
   CeilDiv,
-  And,
+  Lshift,
+  Rshift,
+  Xor,
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type.h:117]: `Xor` should be ~a logical~ an alsoBooleanOperator op.
```
x = torch.randn(10).mul(5).to(dtype=torch.bool) 
y = torch.randn(10).mul(5).to(dtype=torch.bool) 
(x ^ y).dtype # torch.bool
```

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 6]
```diff
@@ -242,28 +227,81 @@ TensorView* arithOpOverloads(
...
+      isIntegralType(v1_dtype) && isIntegralType(v2_dtype);
+
+  if (isIntegerOp(op_type) || (alsoBooleanOperator(op_type) && integer_input)) {
+    // If integer op or maybe bool op with integer inputs meaning binary op
+    if (integer_input && all_integer_input) {
+      return promote_type(v1_dtype, v2_dtype);
+    } else if (integer_input && !all_integer_input) {
+      return isIntegralType(v1_dtype) ? v1_dtype : v2_dtype;
+    } else {
+      return DataType::Int;
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:261]: At this point we have integer operation with inputs that are not integer. Maybe assert false here?

File: test/test_jit_cuda_fuser.py [hunk 7]
```diff
@@ -508,6 +537,50 @@ def test_binary_ops(self):
...
             self._binary_test_helper(op)
 
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_binary_bitwise(self):
+        def jit_or(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
+            return (x & y) | z
+
+        def jit_xor(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:547]: `^` supports boolean inputs. I think this is a case where we would fail.

**Pull request merged**